### DISCUSSION TITLE
Wondering about TUI progress / Keybindings and etc

### DISCUSSION THREAD

**Discussion by 1dancook** [Textualize/rich#1169] [General] [answered]
Hey Will,
I started a project this past week. It would really be best to develop it into a full TUI application. Right now I have stapled on prompt toolkit to highjack key input in order to switch contexts (has a tabbed layout) and do other things. 

I was poking around in the rich repo and saw some TUI things and control things. I couldn't make out with a quick glance what these things will do.

Anyway, I remember a Twitter post about a TUI display. I'm wondering if there's any progress in this area... or even anything I could test? I'd love to ditch prompt toolkit in favor of exclusively using rich. It's been a breeze to build what I have so far using rich. 

Cheers!

**Comment by willmcgugan** [comment, marked as answer]
It's coming. I've made some changes recently under-the-hood to allow for more dyanmic features and I'm making headway with a branch. No promises on timescale, but I'll post progress images / videos on twitter.

**Reply by 1dancook** [reply]
Really looking forward to it! Thanks Will!

**Reply by willmcgugan** [reply]
Work has moved to https://github.com/willmcgugan/textual

**Comment by oleksis** [comment]
> ... I'm wondering if there's any progress in this area... 

Get a look 👀 in the [tui branch](https://github.com/willmcgugan/rich/tree/tui) in this [commit](https://github.com/willmcgugan/rich/commit/f17e51ee124af4036c74c2d225f08e7ba784a5a3)

**Reply by willmcgugan** [reply]
Work in progress! Long way from something usable yet.

**Reply by spanishgum** [reply]
So cool, thank you. I just found this library and started building out my own TUI using `rich.live.Live` + `rich.tree.Tree`.

That event queue is a good idea. I am considering ways I would allow panel switching.

Thinking I would use a prefix command, then an arrow key. That would re-route the "Event" queue to an a handler specific to the active panel.

Super cool library, having lots of fun with it, thanks.

**Reply by willmcgugan** [reply]
The TUI stuff has moved to https://github.com/willmcgugan/textual

**Comment by qTipTip** [comment]
This is very exciting! Loving the `rich` library! 

Is there any low-hanging fruit it is possible to assist with?

**Reply by willmcgugan** [reply]
Rich is quite mature, so there's not many remaining tasks. Although PRs are always appreciated. You're welcome to tackle any outstanding issues.
